mater/lib/src/stores/file.rs

```diff
@@ -0,0 +1,306 @@
+use std::{io::SeekFrom, path::Path};
+
+use indexmap::IndexMap;
+use ipld_core::cid::Cid;
+use sha2::{Digest, Sha256};
+use tokio::{
+    fs::File,
+    io::{AsyncSeekExt, AsyncWriteExt, BufWriter},
+    sync::{Mutex, RwLock},
+};
+
+use crate::{
+    multicodec::SHA_256_CODE,
+    v1::{self, read_block, write_block},
+    v2::{self},
+    CarV1Header, CarV2Header, Characteristics, Error, Index, IndexEntry, MultihashIndexSorted,
+    SingleWidthIndex,
+};
+
+/// Implements a blockstore that stores blocks in a CARv2 format. Blocks put
+/// into the blockstore can be read back once they are successfully written. The
+/// blocks are written immediately, while the index is stored in memory and
+/// updated incrementally.
+///
+/// The blockstore should be closed once the putting blocks is finished. Upon
+/// closing the blockstore, the index is written out to underlying file.
+pub struct FileBlockstore {
+    // Inner store
+    inner: Mutex<FileBlockstoreInner>,
+    // Index of blocks that will be appended to the file at the finalization.
+    // Stored number is an offset that locates the first byte of the block
+    // within the CARv1 payload. The offset is relative to the start of the
+    // CARv1 payload.
+    index: RwLock<IndexMap<Cid, u64>>,
+}
+
+/// Inner file store. Encapsulating state that is locked and used together.
+struct FileBlockstoreInner {
+    // Car file data store
+    store: File,
+    // The byte length of the CARv1 payload. This is used by the indexing, so we
+    // know the locations of each blocks in the file.
+    data_size: u64,
+    // Is true if the blockstore was finalized.
+    is_finalized: bool,
+}
+
+impl FileBlockstore {
+    /// Create a new blockstore. If file at the path already exists it is truncated.
+    pub async fn new<P>(path: P, roots: Vec<Cid>) -> Result<Self, Error>
+    where
+        P: AsRef<Path>,
+    {
```

Comment:
Ideally, you should check that the roots are not empty
```suggestion
    pub async fn new<P>(path: P, roots: Vec<Cid>) -> Result<Self, Error>
    where
        P: AsRef<Path>,
    {
        if roots.is_empty() {
            return Err(crate::Error::EmptyRoots);
        }
```

Reply:
True. Good catch